Reduced memory allocations in the ingesters for the blocks storage

pracucci opened the pull request:
**What this PR does**:
In this PR I'm proposing a solution to reduce memory allocations in the ingesters for the blocks storage, introducing `RefCache` to hold TSDB series references.

Looking at our blocks storage test cluster, we can observe 25% of memory allocations is done by `FromLabelAdaptersToLabelsWithCopy`. Contrary to the chunks storage, whenever we call TSDB `appender.Add()` we have to pass a copy of the labels because they may be retained by TSDB (the request labels are reused to reduce memory allocations).

This PR introduces a series reference cache, so that we can use `appender.AddFast()` instead of `Add()` passing the series reference ID (which is a `uint64`) instead of the labels each time we have the reference cached.

Some notes about the `RefCache` design:
1. It's per-tenant, so we have 1 for each tenant
2. It stripes the cache into 128 buckets to reduce lock contention (128 is a trade-off to not use too much memory on clusters with many tenants but few series, like @thorfour use case)
3. It's intentional having no cap on the max number of series cached in the cache, but there's a purge mechanism to quickly remove stale series. If there would be a cap, once the cap is reached other series would have labels being copied every time, which would allocate that memory anyway until the next GC runs, so keeping all active series references in memory looks a win-win to me.

I'm running this change in our test cluster and memory allocations have been reduced by -25%. In the chart below I'm not showing memory allocation rate > 20 MB/s to hide away the noise every 2 hours when all TSDBs compact the head at the same time (it's another issue, see https://github.com/cortexproject/cortex/issues/2003).

![Screen Shot 2020-01-31 at 08 39 57](https://user-images.githubusercontent.com/1701904/73521400-28030300-4406-11ea-9a6a-aff6b7657d65.png)

I've **not** observed any difference in GC duration time or ingesters response time, neither in the ingesters `go_memstats_heap_inuse_bytes` except that it grows a bit slower because of the reduced allocations, as shown below (same time range):

![Screen Shot 2020-01-31 at 08 50 02](https://user-images.githubusercontent.com/1701904/73521588-b11a3a00-4406-11ea-83a2-07e7fc21b023.png)


**Which issue(s) this PR fixes**:
N/A

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

pstibrany commented on pkg/storage/tsdb/ref_cache.go:
```diff
@@ -0,0 +1,161 @@
+	stripeID := uint8(util.HashFP(fp) % refCacheStripes)
+	stripe := c.stripes[stripeID]
+
+	stripe.refsMu.Lock()
```
The rest of this method can be moved to a method on stripe itself. Also one can use `defer refsMu.Unlock()` here.

pracucci replied:
I didn't use `defer` cause it brings a performance penalty (measured with the benchmark). However, I've moved the logic within a function on the stripe, and locked outside it.

pstibrany commented on pkg/storage/tsdb/ref_cache.go:
```diff
@@ -0,0 +1,161 @@
+
+// Purge removes expired entries from the cache. This function should be called
+// periodically to avoid memory leaks.
+func (c *RefCache) Purge(now int64, ttl int64) {
```
Instead of two values, one can simply pass cut-off time, everything older than that is removed.

codesome replied:
+1 for this.

thorfour commented on pkg/ingester/ingester_v2.go:
```diff
@@ -126,17 +128,22 @@ func NewV2(cfg Config, clientConfig client.Config, limits *validation.Overrides,
 	rateUpdateTicker := time.NewTicker(i.cfg.RateUpdatePeriod)
 	defer rateUpdateTicker.Stop()
 
+	// We use an hardcoded value for this ticket because there should be no
```
`nit` spelling: ticket -> ticker

codesome commented on pkg/storage/tsdb/ref_cache.go:
```diff
@@ -0,0 +1,161 @@
+	stripe := c.stripes[stripeID]
+
+	// Look for series within the stripe
+	stripe.refsMu.Lock()
```
Given all the unlocks are near the returns, we can have it as a `defer` just after this.

pracucci replied:
I didn't use `defer` because there's a performance penalty. According to the benchmark, the benefit of using an array for the stripes (already done) is lost if we introduce a `defer` here.

I've moved the logic to a function to `refCacheStripe` so we can lock outside it, keeping the code cleaner.

pstibrany replied:
> I didn't use `defer` because there's a performance penalty.

I would prefer clearer, more idiomatic code than slightly slower code (which in Go 1.15 will be equally fast anyway).

thorfour replied:
I guess I need to know what the perf penalty actually was otherwise I'd agree with @pstibrany Go 1.14 obviates the need to not defer

pracucci replied:
Ok. Used defer.

thorfour commented on pkg/storage/tsdb/ref_cache.go:
```diff
@@ -0,0 +1,161 @@
+// to copy write request series labels each time (because the memory buffers used to
+// unmarshal the write request is reused).
+type RefCache struct {
+	// The cache is splitted into stripes, each one with a dedicated lock, in
```
`nit` grammer: splitted -> split

codesome commented on pkg/ingester/ingester_v2.go:
```diff
@@ -190,18 +204,45 @@ func (i *Ingester) v2Push(ctx old_ctx.Context, req *client.WriteRequest) (*clien
-		lset := client.FromLabelAdaptersToLabelsWithCopy(ts.Labels)
+		// Check if we already have a cached reference for this series. Be aware
+		// that even if we have a reference it's not guaranteed to be still valid.
+		cachedRef, cachedRefExists := db.refCache.GetRef(now, client.FromLabelAdaptersToLabels(ts.Labels))
```
Can we also comment that `ts.Labels` should be sorted?

thorfour commented on pkg/storage/tsdb/ref_cache.go:
```diff
@@ -0,0 +1,161 @@
+type RefCache struct {
+	// The cache is splitted into stripes, each one with a dedicated lock, in
+	// order to reduce lock contention.
+	stripes map[uint8]*refCacheStripe
```
Since stripes isn't being expanded I think we should use an array (since we have a const) instead of a map.

pracucci replied:
Oh sure! I'm wondering why I've used a map here 🤔

thorfour commented on pkg/storage/tsdb/ref_cache.go:
```diff
@@ -0,0 +1,161 @@
+
+// GetRef returns the cached series reference, and guarantees the input labels set
+// is NOT retained.
+func (c *RefCache) GetRef(now int64, series []labels.Label) (uint64, bool) {
```
`nit` 
`GetRef` should probably just be renamed to `Ref`.
`SetRef` could also be renamed to `Set`

pracucci replied:
Agree for `GetRef` > `Ref`. I think `SetRef` is closer to Go patterns. Am I wrong?

thorfour replied:
yea SetRef can be left as is

thorfour commented on pkg/storage/tsdb/ref_cache.go:
```diff
@@ -0,0 +1,161 @@
+	fp := client.FastFingerprint(client.FromLabelsToLabelAdapters(series))
+
+	// Get the related stripe
+	stripeID := uint8(util.HashFP(fp) % refCacheStripes)
```
related to above about changing from map -> array, since we already hash and modulo we should use an array instead.

thorfour commented on pkg/storage/tsdb/ref_cache.go:
```diff
@@ -0,0 +1,161 @@
+	stripe := c.stripes[stripeID]
+
+	// Look for series within the stripe
+	stripe.refsMu.Lock()
```
probably not worth doing without a profile, but we could likely use a reader lock here, and atomic set the touched at timestamp.

pracucci replied:
I did try it, but the profile didn't show any significative difference, so I've rolled back to this lock which simplify a bit the code.

thorfour replied:
sounds good :+1:

thorfour commented on pkg/storage/tsdb/ref_cache.go:
```diff
@@ -0,0 +1,161 @@
+	}
+}
+
+func (c *RefCache) purgeStripe(now int64, ttl int64, stripe *refCacheStripe) {
```
could squash the `now` and `ttl` params into a single `age` param since all they're used for is to do `now - ttl`

codesome commented on pkg/storage/tsdb/ref_cache_test.go:
```diff
@@ -0,0 +1,113 @@
+	for i := 0; i < b.N; i++ {
+		c.GetRef(0, ls)
+	}
+}
```
Given you are using only 1 series, these benchmarks doesn't seem to be pushing it hard enough.

codesome commented on pkg/ingester/ingester_v2.go:
```diff
@@ -190,18 +204,45 @@ func (i *Ingester) v2Push(ctx old_ctx.Context, req *client.WriteRequest) (*clien
+
+				if err == tsdb.ErrNotFound {
+					cachedRefExists = false
+				}
```
Would be safe to set `err` to `nil` inside here to avoid any subtle bugs in the future.

pracucci commented on pkg/storage/tsdb/ref_cache.go:
```diff
@@ -0,0 +1,161 @@
+
+// GetRef returns the cached series reference, and guarantees the input labels set
+// is NOT retained.
+func (c *RefCache) GetRef(now int64, series []labels.Label) (uint64, bool) {
```
According to the benchmarks, there's a small performance penalty calling `UnixNano()` each time inside the `Get/SetRef()`.

On the other side, `time.Time` is 3x bigger than `uint64`, reason why I didn't store it before. However, for 1M series we're talking about +16MB which is actually negligible, so I've stored `time.Time`.

pstibrany replied:
> According to the benchmarks, there's a small performance penalty calling `UnixNano()` each time inside the `Get/SetRef()`.

This is a weird argument... that penaly must be paid somewhere. In the original code, it was before Get/SetRef was called. I suggested to move that cost inside those method instead. Of course they will be slower, but overall (when taking code *calling* them into account), it’s the same.

pracucci replied:
In the original implementation it was called once for a push request. Moving it inside, it gets called once for each series in the a single push request (ie. if a push request contains 1K series, then it's called 1K times more).

Anyway, I've now stored `time.Time`.

pstibrany replied:
> I suggested to move that cost inside those method instead. Of course they will be slower, but overall (when taking code _calling_ them into account), it’s the same.

I guess the difference is that it’s only called once, when it’s outside of these methods. So what you wrote made sense. My mistake.

pracucci replied:
Yes, if it's called outside it's just called once. No worries! You all left so many good feedback, that the overall change set improved a lot IMO.

pracucci commented on pkg/storage/tsdb/ref_cache.go:
```diff
@@ -0,0 +1,161 @@
+
+// SetRef sets/updates the cached series reference. The input labels set IS retained.
+func (c *RefCache) SetRef(now int64, series []labels.Label, ref uint64) {
+	fp := client.FastFingerprint(client.FromLabelsToLabelAdapters(series))
```
I initially picked `FastFingerprint()` because it's supposed by the "fast", but I've run the benchmark with `Fingerprint()` and the penalty is negligible.

pracucci commented on pkg/storage/tsdb/ref_cache.go:
```diff
@@ -0,0 +1,161 @@
+
+// refCacheStripe holds a subset of the series references for a single tenant.
+type refCacheStripe struct {
+	refsMu sync.Mutex
```
Unfortunately not because the getter also touch the timestamp.

codesome replied:
Ah yes. If we face a lot of contention for any reason in the future, we can try our hands on atomics as a follow on work.

pracucci commented:
@codesome @pstibrany @thorfour Thanks for the feedback. I should have addressed all comments. May you check it again, please?

codesome commented on pkg/storage/tsdb/ref_cache.go:
```diff
@@ -0,0 +1,157 @@
+	// DefaultRefCacheTTL is the default RefCache purge TTL. We use a reasonable
+	// value that should cover most use cases. The cache would be ineffective if
+	// the scrape interval of a series is greater than this TTL.
+	DefaultRefCacheTTL = int64(5 * time.Minute)
```
Why not `time.Duration(5 * time.Minute)` as it's only used that way?

pracucci replied:
Right, done. Was incorrectly left as `int64` by the initial implementation.

gouthamve commented on pkg/storage/tsdb/ref_cache.go:
```diff
@@ -0,0 +1,157 @@
+	// DefaultRefCacheTTL is the default RefCache purge TTL. We use a reasonable
+	// value that should cover most use cases. The cache would be ineffective if
+	// the scrape interval of a series is greater than this TTL.
+	DefaultRefCacheTTL = 5 * time.Minute
```
Can we make this `10 * time.Minute`? `5min` is still a common scrape interval on some intensive targets.

gouthamve commented on pkg/storage/tsdb/ref_cache.go:
```diff
@@ -0,0 +1,157 @@
+	// the scrape interval of a series is greater than this TTL.
+	DefaultRefCacheTTL = 5 * time.Minute
+
+	refCacheStripes = 128
```
`refCacheStripes` --> `numRefCacheStripes`

Down in L29, it looks like an array of `refCacheStripe`.

pracucci commented:
Thanks @gouthamve for your review. I've addressed the comments.

> Though I must admit, I am a little worried that the purge process might block ingestion. I'd rather see it happen before we optimise for that though.

It's a legit objection. I've added a benchmark to measure it. Given 1M series in the cache and 10K expired ones, on my MacBook it takes 400ms. We have 128 stripes (the lock is individually taken for each stripe), so each lock is acquired (on average) 3.2ms. Doesn't look dramatic, but if we see performances issues because of this we can add optimizations.
